### DISCUSSION TITLE
tutorials/lift-off-part2/05-the-shape-of-a-resolver

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#48] [course feedback]
# tutorials/lift-off-part2/05-the-shape-of-a-resolver

https://www.apollographql.com/tutorials/lift-off-part2/05-the-shape-of-a-resolver

**Comment by vasily-mishanin** [comment]
Should we add the code from "Code Challenge!" into the course project's code?

**Reply by lizhennessy** [reply]
Hi @vasily-mishanin! The Code Challenge is just an exercise space for you to practice what you learned in this lesson. You do not need to add any of its content into your course project; you'll be able to tackle that directly in the next lesson. Please let us know if you have any other questions!

**Reply by anvodev** [reply]
In addition I think we should do the code challenge after we finish the source code so we can do the challenge confidently without looking for solution

**Comment by musatcristian** [comment]
Hi, 
Love the course,
I have a small question regarding implemeting a resolver in  typescript: 
  Is there a type for a resolver function ? that could help with autocomplete and others ?
  e.g: `export const tracksForHomeResolver: ` __what type can I put here ???__`= ({}, {}, {`__so I get autocomplete here__`}, {}) => {...};`

**Reply by musatcristian** [reply]
nevermind the above comment.
I got to the codegen lesson.
you guys are awesome

**Reply by lizhennessy** [reply]
Awesome, @musatcristian ! So glad to hear that the pieces are coming together. Please don't hesitate to reach out if you have any other questions.

**Comment by marksikaundi** [comment]
This course is so interesting, if you guys can create more contents like this...definately alot will. learn

**Comment by belloshehu** [comment]
Looking at the way resolvers return values for fields in query from difference data sources while following schema, I don't know if I will be correct to say that they are the magic behind GraphQL's flexibility to return data in various shapes and depth.

**Reply by Theedon** [reply]
While they don't define the shape of the data; they are in charge of retrieving it and making it into the proper shape. So yes, they definitely are a big part of the GraphQL magic.

**Comment by ponna93** [comment]
Hi All, can some one point out how to do the same in java springboot please ?

**Reply by lizhennessy** [reply]
Hi there @ponna93 ! We have another course specifically for Java, Spring Boot, and
the Domain Graph Service (DGS) framework. Here's the link!
https://www.apollographql.com/tutorials/intro-dgs
